File: src/lieb_liniger_state.cpp
#include "constants.hpp"
#include "lieb_liniger_state.hpp"
#include "machine_learning.hpp"

#include <Eigen/Dense>

#include <algorithm>
#include <iomanip>
#include <iostream>
#include <numeric>
#include <random>

/**
 * Construct a Bethe state with Bethe numbers of the ground state.
 */
lieb_liniger_state::lieb_liniger_state(double new_c, double new_L,
                                       int new_N)
    : c(new_c),
      L(new_L),
      N(new_N),
      Is(new_N),
      lambdas(new_N),
      gaudin_matrix(new_N, new_N)
{
    generate_gs_bethe_numbers();

}

/**
 * Construct a Bethe state with arbitrary Bethe numbers.
 */
lieb_liniger_state::lieb_liniger_state(double new_c, double new_L,
                                       int new_N,
                                       Eigen::VectorXd new_Is)
    : c(new_c),
      L(new_L),
      N(new_N),
      Is(new_Is),
      lambdas(new_N),
      gaudin_matrix(new_N, new_N)
{
}

/**
 * Generate the Bethe numbers of the N particle Lieb-Liniger ground state.
 */
void lieb_liniger_state::generate_gs_bethe_numbers() {
    if (N % 2 == 0) {
        Is = Eigen::VectorXd::LinSpaced(N, 1, N) - Eigen::VectorXd::Constant(N,  N/2);
        Is -= Eigen::VectorXd::Constant(N, 0.5);
    } else {
        Is = Eigen::VectorXd::LinSpaced(N, -(N-1)/2, (N-1)/2);
    }
}

void lieb_liniger_state::find_rapidities(bool use_machine_learning) {
    if (!use_machine_learning) {
        // Initial bad guess for the rapidities.
        lambdas = 2 * PI / L * Is;
        calculate_rapidities_newton();
    } else {
        lambdas = guess_rapidities(Is);
        calculate_rapidities_newton();
        update_neural_net(Is, lambdas);
    }
}

/**
 * Calculate the rapidities of the N particle Lieb-Liniger state.
 *
 * We use a multi-dimensional Newton method, where for a function F
 * we have \f$J_F(x_n) (x_{n+1} - x_n) = - F(x_n)\f$, with \f$J_F\f$ the 
 * Jacobian, which in this case is the Gaudin matrix.
 *
 * This function is largely inspired by the Find_Rapidities() function
 * for the Lieb-Liniger model found in the ABACUS library by J-S Caux. 
 */
void lieb_liniger_state::calculate_rapidities_newton() {
    for (int no_of_iterations = 0; no_of_iterations < 20; no_of_iterations++) {
        // Calculate the Yang-Yang equation values.
        Eigen::VectorXd rhs_bethe_equations = Eigen::VectorXd(N).setZero();
        for (int j = 0; j < N; j++) {
            for (int k = 0; k < N; k++) {
                rhs_bethe_equations(j) += 2 * atan((lambdas(j) - lambdas(k) / c));
            }
            rhs_bethe_equations(j) += L * lambdas(j) - 2 * PI * Is(j);
        }

        // Perform a step of the Newton method.
        calculate_gaudin_matrix();
        // Partial LU decomposition may be used sinde the Gaudin matrix is non-singular.
        Eigen::VectorXd delta_lambda = gaudin_matrix.lu().solve(-rhs_bethe_equations);

        // Calculate the average difference squared of the rapidity changes.
        double diff_square = 0;
        for (int i = 0; i < N; i++) {
            diff_square += delta_lambda(i) * delta_lambda(i) / lambdas(i) * lambdas(i);
        }
        diff_square /= N;

        // Update rapidities.
        for (int p = 0; p < N; p++) {
            lambdas(p) += delta_lambda(p);
        }

        // Rapidities converged.
        if (diff_square < 10e4 * MACHINE_EPS_SQUARE) {
            break;
        }
    }
}

/**
 * Calculate the Gaudin matrix of a Lieb-Liniger state.
 */
void lieb_liniger_state::calculate_gaudin_matrix() {
    for (int j = 0; j < lambdas.size(); j++) {
        for (int k = 0; k < lambdas.size(); k++) {
            if (j == k) {
                double kernel_sum = L;
                for (int kp = 0; kp < N; kp++) {
                    kernel_sum += kernel(lambdas(j) - lambdas(kp), c);
                }
                gaudin_matrix(j, k) = kernel_sum;
            } else {
                gaudin_matrix(j, k) = -kernel(lambdas(j) - lambdas(k), c);
            }
        }
    }
}

/**
 * Calculate the kernel of the Lieb-Liniger model.
 */
double lieb_liniger_state::kernel(double k, double c) {
    return 2 * c / (c * c + k * k);
}

/**
 * Generate a vector of Bethe numbers distributed according to a
 * Gaussian distribution, with only unique entries.
 */
Eigen::VectorXd generate_bethe_numbers(int N) {
    Eigen::VectorXd bethe_numbers = Eigen::VectorXd::Constant(N, -10e7);
    double mean = 0;
    double standard_dev = N * PI; // We assume that excitations still cluster around the ground state.

    std::random_device rd;
    std::mt19937 gen(rd());
 
    std::normal_distribution<> normal_distribution(mean, standard_dev);
    std::uniform_real_distribution<> sign_distribution(-1, 1);

    // TODO: Make sure all entries are unique.
    for (int n = 0; n < N; n++) {
        double random_number = std::round(normal_distribution(gen));
        // std::cout << "rand num " << random_number << std::endl;

        // For N even Bethe numbers are half odd.
        if (N % 2 == 0) {
            bethe_numbers(n) = random_number + std::copysign(1, sign_distribution(gen)) * 0.5;            
        } else {
            bethe_numbers(n) = random_number;
        }
    }

    std::sort(bethe_numbers.data(), bethe_numbers.data() + bethe_numbers.size());
    
    return bethe_numbers;
}

File: python/lieb_liniger_state.py
import numpy as np


class lieb_liniger_state:
    def __init__(self, c, L, N, bethe_numbers=None):
        self.c = c
        self.L = L
        self.N = N
        self.energy = 0
        self.momentum = 0
        self.integer_momentum = 0
        self.norm = np.inf
        self.ff = 0
        if bethe_numbers is not None:
            self.Is = bethe_numbers
        else:
            self.Is = self.generate_gs_bethe_numbers()
        self.lambdas = np.zeros(N)
        self.gaudin_matrix = np.zeros((self.N, self.N))

    def generate_gs_bethe_numbers(self):
        """Generate ground state Bethe numbers for the Lieb-Linger model."""
        if self.N % 2 == 0:
            return -0.5 - self.N/2 + np.linspace(1, self.N, self.N)
        else:
            return np.linspace(-(self.N-1)/2, (self.N-1)/2, self.N)

    @staticmethod
    def kernel(k, c):
        return 2 * c / (c**2 + k**2)

    def calculate_gaudin_matrix(self):
        # """Calculate the Gaudin matrix."""
        for j in range(self.N):
            for k in range(self.N):
                if j == k:
                    kernel_sum = self.L
                    for kp in range(self.N):
                        kernel_sum += self.kernel(self.lambdas[j] - self.lambdas[kp], self.c)
                    self.gaudin_matrix[j, k] = kernel_sum - self.kernel(0, self.c)
                else:
                    self.gaudin_matrix[j, k] = -self.kernel(self.lambdas[j] - self.lambdas[k], self.c)

    def calculate_rapidities_newton(self, printing=False):
        """Calculate the rapidities using a multidimensional Newton method."""
        # TODO: Implement relaxation as done by Caux.
        for no_of_iterations in range(20):
            rhs_bethe_equations = np.zeros(self.N)
            for j in range(self.N):
                for k in range(self.N):
                    rhs_bethe_equations[j] += 2 * np.arctan((self.lambdas[j] - self.lambdas[k]) / self.c)
                rhs_bethe_equations[j] += self.L * self.lambdas[j] - 2 * np.pi * self.Is[j]

            self.calculate_gaudin_matrix()
            delta_lambda = np.linalg.solve(self.gaudin_matrix, -rhs_bethe_equations)

            diff_square = 0
            for i in range(self.N):
                diff_square += delta_lambda[i]**2 / (self.lambdas[i]**2 + 10**-6)
            diff_square /= self.N
            if printing:
                print(diff_square)

            for i in range(self.N):
                self.lambdas[i] += delta_lambda[i]

            if diff_square < 10**-14:
                return no_of_iterations

        return no_of_iterations

    def calculate_rapidities(self):
        self.lambdas = 2 * np.pi / self.L * np.array(self.Is)
        self.calculate_rapidities_newton()

    def calculate_energy(self):
        self.energy = np.sum(self.lambdas**2)

    def calculate_momentum(self):
        self.momentum = np.sum(self.lambdas)

    def calculate_integer_momentum(self):
        self.integer_momentum = np.sum(self.Is)

    def calculate_norm(self):
        self.norm = self.c**self.N * np.linalg.det(self.gaudin_matrix)
        for k in range(self.N):
            for j in range(k+1, self.N):
                self.norm *= ((self.lambdas[j] - self.lambdas[k])**2 + self.c**2) / (self.lambdas[j] - self.lambdas[k])**2

    def calculate_all(self):
        self.calculate_rapidities()
        self.calculate_energy()
        self.calculate_momentum()
        self.calculate_integer_momentum()
        self.calculate_norm()


def generate_bethe_numbers(N, ref_state, max_I = np.inf):
    """Generate Bethe numbers for excited states."""
    bethe_numbers = np.full(N, 10.**7, dtype=np.float)
    no_of_unique_entries = 0
    while no_of_unique_entries < N:
        random_number = np.round(np.random.normal(0, N))
        if N % 2 == 0:
            random_number += np.sign(np.random.uniform(-1, 1)) * 0.5

        if (np.abs(bethe_numbers - random_number) < 10e-4).any():
            continue
        # Implement a UV cutoff in Bethe number space.
        elif abs(random_number) > max_I:
            continue
        else:
            bethe_numbers[no_of_unique_entries] = random_number
            no_of_unique_entries += 1

    # Make sure we do not generate the reference state.
    if list(np.sort(bethe_numbers)) == ref_state:
        return generate_bethe_numbers(N, ref_state)
    else:
        return np.sort(bethe_numbers)


def mutate_bethe_numbers(bethe_numbers):
    """Mutate a given state of Bethe numbers to have an excited particle."""
    while True:
        rand_index = np.random.randint(len(bethe_numbers))
        new_bethe_numbers = list(bethe_numbers[:rand_index]) + list(bethe_numbers[rand_index+1:])
        new_bethe_numbers += [bethe_numbers[rand_index] + np.sign(np.random.uniform(-1, 1))]
        if len(np.unique(new_bethe_numbers)) == len(bethe_numbers):
            return sorted(new_bethe_numbers)


if __name__ == "__main__":
    with open("lieblinigerc1L100N10random.txt", "a+") as file:
        for i in range(10000):
            bethe_numbers = generate_bethe_numbers(10)
            llstate = lieb_liniger_state(1, 100, 10, bethe_numbers)
            llstate.calculate_all()
            file.write(str(repr(list(llstate.Is))) + "\n" + str(llstate.energy) + "\n")
